src/main/java/org/openrewrite/java/migrate/util/OptionalStreamRecipe.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2023 Atlassian US Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.migrate.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.tree.Comment;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JRightPadded;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Space;
+import org.openrewrite.java.tree.TextComment;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class OptionalStreamRecipe extends Recipe {
+  @Override
+  public String getDisplayName() {
+    return "Stream<Optional> idiom recipe";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Migrate Java 8 optional stream idiom .filter(Optional::isPresent).map(Optional::get) to Java 11 .flatMap(Optional::stream).";
+  }
+
+  @Override
+  public TreeVisitor<?, ExecutionContext> getVisitor() {
+    return new OptionalStreamVisitor();
```

Comment:
Probably best to add a quick limiting `Precondition.check` here to only process classes that use the offending classes and methods, for typically better performance. I'd think that `new UsesMethod("java.lang.Optional isPresent()")` could work well here, as that's the one method that's infrequently used, and an essential part of the pattern we want to replace here.

Reply:
After I added
`return Preconditions.check(new UsesMethod<>("java.lang.Optional isPresent()"), new OptionalStreamVisitor());`
tests fail with
```
java.lang.AssertionError: Expected recipe to complete in 1 cycle, but took 0 cycle.
	at org.openrewrite.test.LargeSourceSetCheckingExpectedCycles.afterCycle(LargeSourceSetCheckingExpectedCycles.java:80)
	at org.openrewrite.RecipeScheduler.scheduleRun(RecipeScheduler.java:88)
	at org.openrewrite.Recipe.run(Recipe.java:279)
        ......
```

Reply:
Sure! I'm traveling right now, but will try to pick this up Friday; unless @joanvr beats me to it.

Reply:
Applied just now :)